odoo/orm/fields_relational.py

```diff
@@ -459,51 +459,62 @@ def condition_to_sql(self, field_expr: str, operator: str, value, model: BaseMod
             # for other operators than 'any', just generate condition based on column type
             return super().condition_to_sql(field_expr, operator, value, model, alias, query)
 
-        fname = field_expr
         comodel = model.env[self.comodel_name]
-        sql_field = model._field_to_sql(alias, fname, query)
+        sql_field = model._field_to_sql(alias, field_expr, query)
         can_be_null = self not in model.env.registry.not_null_fields
-
-        if not isinstance(value, Domain):
-            # value is SQL or Query
-            if isinstance(value, Query):
-                subselect = value.subselect()
-            elif isinstance(value, SQL):
-                subselect = SQL("(%s)", value)
-            else:
-                raise TypeError(f"condition_to_sql() 'any' operator accepts Domain, SQL or Query, got {value}")
-            sql = SQL(
-                "%s%s%s",
-                sql_field,
-                SQL(" IN ") if operator in ('any', 'any!') else SQL(" NOT IN "),
-                subselect,
+        bypass_access = operator in ('any!', 'not any!') or self.bypass_search_access
+        positive = operator in ('any', 'any!')
+
+        # Decide whether to use a LEFT JOIN
+        left_join = bypass_access and isinstance(value, Domain)
+        if left_join and not positive:
+            # For 'not any!', we get a better query with a NOT IN when we have a
+            # lot of positive conditions which have a better chance to use
+            # indexes.
+            #   `field NOT IN (SELECT ... WHERE z = y)` better than
+            #   `LEFT JOIN ... ON field = id WHERE z <> y`
+            # There are some exceptions: we filter on 'id'.
+            left_join = sum(
+                (-1 if Domain.is_negative_operator(cond.operator) else 1)
+                for cond in value.iter_conditions()
+            ) < 0 or any(
+                cond.field_expr == 'id' and not Domain.is_negative_operator(cond.operator)
+                for cond in value.iter_conditions()
             )
-            if can_be_null and operator in ('not any', 'not any!'):
-                sql = SQL("(%s IS NULL OR %s)", sql_field, sql)
-            if self.company_dependent:
-                sql = self._condition_to_sql_company(sql, field_expr, operator, value, model, alias, query)
-            return sql
-
-        # value is a Domain
 
-        if self.bypass_search_access or operator in ('any!', 'not any!'):
+        if left_join:
             comodel, coalias = self.join(model, alias, query)
-
+            if not positive:
+                value = (~value).optimize_full(model.sudo())
```

Comment:
Optimizing with `model.sudo()` does not look correct to me, since it will also optimize the domain of "any" conditions in `~value`.  In particular if `any` conditions are transformed into `any!` conditions...